mantaray/received.py

```diff
@@ -316,14 +335,35 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
     subcommand = args[1]
     if subcommand == "ACK":
         acknowledged = set(args[-1].split())
+
         if "sasl" in acknowledged:
             server_view.core.send("AUTHENTICATE PLAIN")
+
+        for capability in acknowledged:
+            server_view.core.cap_list.add(capability)
+
     elif subcommand == "NAK":
         rejected = set(args[-1].split())
         if "sasl" in rejected:
             # TODO: this good?
             raise ValueError("The server does not support SASL.")
 
+    else:
+        server_view.core.send("CAP END")
+        raise ValueError("Invalid CAP response. Aborting Capability Negotiation.")
+
+    server_view.core.pending_cap_count -= 1
+
+    # If CAP END is sent before the SASL authentication message (_handle_authenticate()),
+    # the server will respond with 904 SASL authentication aborted.
+    # If SASL has been negotiated with the server, Mantaray sends CAP END after the server
+    # has replied with an authentication success/fail message
```

Comment:
I still have a hard time understanding this comment. How about this? it includes more things that you can easily `git grep`.

```suggestion
    # If "sasl" is in cap_list, Mantaray sends CAP END after the server
    # has replied with RPL_SASLSUCCESS or ERR_SASLFAIL
```

Reply:
I thought you wanted docstrings/comments to be more high level. That's why I wrote it like this. High level maybe only applies to docstrings?